tdrs-backend/tdpservice/parsers/validators.py

```diff
@@ -61,16 +127,159 @@ def startsWith(substring):
+        lambda value: f'{value} is not smaller than {UpperBound}.'
+    )
+
+def isInLimits(LowerBound, UpperBound):
+    """Validate that value is in a range including the limits."""
+    return make_validator(
+        lambda value: value >= LowerBound and value <= UpperBound,
+        lambda value: f'{value} is not larger and equal to {LowerBound} and smaller and equal to {UpperBound}.'
     )
 
 
 # custom validators
 
+def month_year_monthIsValid():
+    """Validate that in a monthyear combination, the month is a valid month."""
+    return make_validator(
+        lambda value: int(str(value)[4:6]) in range(1, 13),
+        lambda value: f'{str(value)[4:6]} is not a valid month.'
+    )
+
+
+def month_year_yearIsLargerThan(year):
+    """Validate that in a monthyear combination, the year is larger than the given year."""
+    return make_validator(
+        lambda value: int(str(value)[:4]) > year,
+        lambda value: f'{str(value)[:4]} year must be larger than {year}.'
+    )
+
+# outlier validators
+def validate__FAM_AFF__SSN():
+    """If item 30 ==2 and item 42 ==1 or 2, then item 33 != 000000000 -- 999999999."""
+    # value is instance
+    def validate(instance):
+        FAMILY_AFFILIATION = instance['FAMILY_AFFILIATION'] if type(instance) is dict else \
+            getattr(instance, 'FAMILY_AFFILIATION')
+        CITIZENSHIP_STATUS = instance['CITIZENSHIP_STATUS'] if type(instance) is dict else \
+            getattr(instance, 'CITIZENSHIP_STATUS')
+        SSN = instance['SSN'] if type(instance) is dict else getattr(instance, 'SSN')
+        if FAMILY_AFFILIATION == 2 and (CITIZENSHIP_STATUS == 1 or CITIZENSHIP_STATUS == 2):
+            if SSN in [str(i) * 9 for i in range(10)]:
+                return (False,
+                        'If FAMILY_AFFILIATION ==2 and CITIZENSHIP_STATUS==1 or 2, then SSN != 000000000 -- 999999999.')
+            else:
+                return (True, None)
+        else:
+            return (True, None)
+    return lambda instance: validate(instance)
+
+def validate__FAM_AFF__HOH__Fed_Time():
+    """If FAMILY_AFFILIATION == 2 and MONTHS_FED_TIME_LIMIT== 1 or 2, then MONTHS_FED_TIME_LIMIT > 1."""
+    # value is instance
+    def validate(instance):
+        FAMILY_AFFILIATION = instance['FAMILY_AFFILIATION'] if type(instance) is dict else \
+            getattr(instance, 'FAMILY_AFFILIATION')
+        RELATIONSHIP_HOH = instance['RELATIONSHIP_HOH'] if type(instance) is dict else \
+            getattr(instance, 'RELATIONSHIP_HOH')
+        MONTHS_FED_TIME_LIMIT = instance['MONTHS_FED_TIME_LIMIT'] if type(instance) is dict else \
+            getattr(instance, 'MONTHS_FED_TIME_LIMIT')
+        if FAMILY_AFFILIATION == 1 and (RELATIONSHIP_HOH == 1 or RELATIONSHIP_HOH == 2):
+            if int(MONTHS_FED_TIME_LIMIT) < 1:
+                return (False,
+                        'If FAMILY_AFFILIATION == 2 and MONTHS_FED_TIME_LIMIT== 1 or 2, then MONTHS_FED_TIME_LIMIT > 1.'
```

Comment:
what is this?

Reply:
it is this: If item 30 == 1 and item 38 == 1 or 2, then item 44 >= 001